rfc/src/rfcs/0008-source-coverage.md

```diff
@@ -0,0 +1,300 @@
+
+Therefore, the first part of the coverage workflow will be managed by Kani.
+Then, in the other part, we will use the `kani-cov` tool to produce the coverage
+output(s) we're interested in.
+
+In the following, we describe each one of these components in more detail.
+
+#### The `-cov` option
+
+The coverage workflow will be kicked off through a new `-cov` option:
+
+```sh
+cargo kani -cov
+```
+
+The main difference with respect to the regular verification workflow is that,
+at the end of the verification-based coverage run, Kani will generate two types
+of files:
+ - One single file `.kanimap` file for the project. This file will contain the
+ coverage mappings for the project's source code.
+ - One `.kaniraw` file for each harness. This file will contain the
+ verification-based results for the coverage-oriented properties corresponding
+ to a given harness.
+
+Note that `.kaniraw` files correspond to `.profraw` files in the LLVM coverage
+workflow. Similarly, the `.kanimap` file corresponds to the coverage-related
+information that's embedded into the project's binaries in the LLVM coverage
+workflow.[^note-kanimap]
+
+The files will be written into a new timestamped directory associated with the
+coverage run. The path to this directory will be printed to standard output in
+by default. For example, the [draft implementation](https://github.com/model-checking/kani/pull/3119)
+writes the coverage files into the `target/kani/<target_triple>/cov/` directory.
+
+Users aren't expected to read the information in any of these files.
+Therefore, there's no need to restrict their format.
+The [draft implementation](https://github.com/model-checking/kani/pull/3119)
+uses the JSON format but we might consider using a binary format if it doesn't
+scale.
+
+[^note-kanimap]: Note that the `.kanimap` generation isn't implemented in [#3119](https://github.com/model-checking/kani/pull/3119).
+The [draft implementation of `kani-cov`](https://github.com/model-checking/kani/pull/3121)
+simply reads the source files referred to by the code coverage checks, but it
+doesn't get information about code trimmed out by the MIR linker.
+
+#### The `kani-cov` tool
+
+The `kani-cov` tool will be used to process coverage information generated by
+Kani and produce coverage outputs as indicated by the user.
+Hence, the `kani-cov` tool corresponds to the set of LLVM tools
+(`llvm-profdata`, `llvm-cov`, etc.) that are used to produce coverage outputs
+through the LLVM coverage workflow.
+
+In contrast to LLVM, we'll have a single tool for all Kani coverage-related needs.
+We suggest that the tool initially offers three subcommands[^note-export]:
+ - `merge`: to combine the coverage results of one or more `.kaniraw` files into
+ a single `.kanicov` file, which will be required for the other subcommands.
+ - `report`: to display a summary of the coverage results.
+ - `show`: to produce source-based code coverage reports in human-readable formats (e.g., HTML).
```

Comment:
Do we need this? I'm wondering if we really need to split Kani's coverage flow into multiple commands.

It makes sense to me that LLVM does that, since users want to execute the binaries independently and merge the results after the fact. But with Kani, `kani-driver` is already responsible for compiling and running the verification, so could `kani-driver` also aggregate the results at the end?

BTW, I'm OK if we want to put that logic in its own separate binary (`kani-cov`), but I'm wondering if `kani-driver` should be the one invoking `kani-cov`.

Reply:
We don't really need this, but I think there two main reasons to propose it:
 1. **To resemble the LLVM workflow as much as possible**, potentially easing an integration with the LLVM framework in the future. Unfortunately, the integration with LLVM tooling isn't completely explained in the RFC at the moment, but I think you get the idea.
 2. **To allow users to aggregate and extract coverage data however they want** (i.e., to have flexible coverage options).

But you're right in that this may not represent the best experience for Kani users. The three sub-commands provide much flexibility, yes. But based on my own experience, they aren't that ergonomic and I'm always copy-pasting the same ones most of the time.

Reply:
Maybe we can have a unified flow out of the box, and later allow users to aggregate them in different ways

Reply:
> Maybe we can have a unified flow out of the box, and later allow users to aggregate them in different ways

Yes, I'm going to make changes to the RFC and propose something like that. Thanks!